File: android/src/main/kotlin/com/monorithm/monolens/ImageTransformer.kt
package com.monorithm.monolens

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import androidx.exifinterface.media.ExifInterface
import java.io.File
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Crop, rotate, flip, downscale and re-encode a still.
 *
 * One decode and one bitmap op. The output size is worked out first, from
 * bounds metadata alone, and the decode then subsamples to just past what that
 * output needs — cropping a 12 MP photo to a 1080 px square never allocates the
 * 12 MP bitmap, which is also what keeps this off the OOM killer's radar.
 * `createBitmap` takes a sub-rectangle *and* a matrix, so the crop, rotation,
 * mirror and final scale all happen in that single call rather than in the
 * three chained allocations this used to do.
 */
internal object ImageTransformer {

    private data class Rect(val x: Int, val y: Int, val width: Int, val height: Int)

    fun apply(request: ImageEditRequest): MediaInfo {
        if (!File(request.sourcePath).exists()) {
            throw MonolensException.notFound(request.sourcePath)
        }

        val orientation = orientationOf(request.sourcePath)
        val full = uprightSize(request.sourcePath, orientation)
        val crop = cropRect(request.crop, full)
        val output = outputSize(crop, request.rotation, request.maxDimension)

        // Whole frame : kept region :: what we decode : output.
        val cropLongest = max(crop.width, crop.height)
        val fullLongest = max(full.width, full.height)
        val wanted = if (cropLongest <= 0) {
            fullLongest
        } else {
            ceil(
                max(output.width, output.height).toDouble() * fullLongest / cropLongest,
            ).toInt()
        }

        var bitmap = decode(request.sourcePath, sampleSize(fullLongest, wanted))
        try {
            // Only when the tag says so — most captures are already upright and
            // skip this pass entirely.
            bitmap = bitmap.uprighted(orientation)
            bitmap = bitmap.transformed(crop, full, output, request)

            if (request.annotations.isNotEmpty()) {
                // Blur first: it samples the picture, so it belongs to it. Text,
                // stickers and strokes then go over the result, which is what
                // keeps a caption across a blurred face legible.
                bitmap = AnnotationRenderer.applyBlur(bitmap, request.annotations)
                bitmap = bitmap.withOverlay(
                    AnnotationRenderer.overlayLayer(
                        request.annotations,
                        bitmap.width,
                        bitmap.height,
                    ),
                )
            }

            val format = when (request.format) {
                MonoImageFormat.PNG -> Bitmap.CompressFormat.PNG
                MonoImageFormat.JPEG -> Bitmap.CompressFormat.JPEG
            }
            val file = File(request.outputPath)
            file.parentFile?.mkdirs()
            file.outputStream().use { stream ->
                if (!bitmap.compress(format, request.quality.toInt().coerceIn(1, 100), stream)) {
                    throw MonolensException.encodeFailed("Could not encode to ${request.format}")
                }
            }

            return MediaInfo(
                path = request.outputPath,
                width = bitmap.width.toLong(),
                height = bitmap.height.toLong(),
                durationMs = null,
                byteSize = file.length(),
            )
        } finally {
            bitmap.recycle()
        }
    }

    // Geometry — all in full-resolution upright pixels, so the output dimensions
    // are exact and independent of how far the decoder subsampled.

    private fun orientationOf(path: String): Int = runCatching {
        ExifInterface(path).getAttributeInt(
            ExifInterface.TAG_ORIENTATION,
            ExifInterface.ORIENTATION_NORMAL,
        )
    }.getOrDefault(ExifInterface.ORIENTATION_NORMAL)

    private fun uprightSize(path: String, orientation: Int): Rect {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(path, options)
        if (options.outWidth <= 0 || options.outHeight <= 0) {
            throw MonolensException.decodeFailed("Could not read $path")
        }
        val turned = orientation == ExifInterface.ORIENTATION_ROTATE_90 ||
            orientation == ExifInterface.ORIENTATION_ROTATE_270 ||
            orientation == ExifInterface.ORIENTATION_TRANSPOSE ||
            orientation == ExifInterface.ORIENTATION_TRANSVERSE
        return Rect(
            0,
            0,
            if (turned) options.outHeight else options.outWidth,
            if (turned) options.outWidth else options.outHeight,
        )
    }

    private fun cropRect(rect: NormalizedRect?, full: Rect): Rect {
        if (rect == null) return full
        val x = (rect.left * full.width).roundToInt().coerceIn(0, full.width - 1)
        val y = (rect.top * full.height).roundToInt().coerceIn(0, full.height - 1)
        return Rect(
            x,
            y,
            (rect.width * full.width).roundToInt().coerceIn(1, full.width - x),
            (rect.height * full.height).roundToInt().coerceIn(1, full.height - y),
        )
    }

    private fun outputSize(crop: Rect, rotation: MonoRotation, cap: Long?): Rect {
        val turned = rotation == MonoRotation.QUARTER_TURN ||
            rotation == MonoRotation.THREE_QUARTER_TURNS
        var width = if (turned) crop.height else crop.width
        var height = if (turned) crop.width else crop.height

        val limit = cap?.toInt()
        if (limit != null && limit >= 1 && max(width, height) > limit) {
            val factor = limit.toDouble() / max(width, height)
            width = (width * factor).roundToInt().coerceAtLeast(1)
            height = (height * factor).roundToInt().coerceAtLeast(1)
        }
        return Rect(0, 0, width, height)
    }

    /** The largest power-of-two subsample that still leaves [wanted] pixels. */
    private fun sampleSize(fullLongest: Int, wanted: Int): Int {
        if (wanted <= 0) return 1
        var sample = 1
        while (fullLongest / (sample * 2) >= wanted) sample *= 2
        return sample
    }

    // Pixels

    private fun decode(path: String, sampleSize: Int): Bitmap {
        val options = BitmapFactory.Options().apply {
            inSampleSize = sampleSize
            // The transform below reads every pixel exactly once, so there is
            // nothing to gain from a mutable or higher-precision config.
            inPreferredConfig = Bitmap.Config.ARGB_8888
        }
        return BitmapFactory.decodeFile(path, options)
            ?: throw MonolensException.decodeFailed("Could not decode $path")
    }

    /** Draws [layer] over this bitmap, making it mutable first if it is not. */
    private fun Bitmap.withOverlay(layer: Bitmap?): Bitmap {
        if (layer == null) return this
        val target = if (isMutable) this else copy(Bitmap.Config.ARGB_8888, true)
        if (target !== this) recycle()
        Canvas(target).drawBitmap(layer, 0f, 0f, null)
        layer.recycle()
        return target
    }

    private fun Bitmap.uprighted(orientation: Int): Bitmap {
        val matrix = Matrix()
        when (orientation) {
            ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
            ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
            ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
            ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
            ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
            ExifInterface.ORIENTATION_TRANSPOSE -> {
                matrix.postRotate(90f)
                matrix.postScale(-1f, 1f)
            }

            ExifInterface.ORIENTATION_TRANSVERSE -> {
                matrix.postRotate(270f)
                matrix.postScale(-1f, 1f)
            }

            else -> return this
        }
        val result = Bitmap.createBitmap(this, 0, 0, width, height, matrix, true)
        if (result != this) recycle()
        return result
    }

    /** Crop, scale, rotate and mirror in one allocation. */
    private fun Bitmap.transformed(
        crop: Rect,
        full: Rect,
        output: Rect,
        request: ImageEditRequest,
    ): Bitmap {
        // The decode was approximate, so map the crop into whatever came back
        // rather than assuming it honoured the request exactly.
        val scaleX = width.toDouble() / full.width
        val scaleY = height.toDouble() / full.height
        val x = (crop.x * scaleX).toInt().coerceIn(0, width - 1)
        val y = (crop.y * scaleY).toInt().coerceIn(0, height - 1)
        val w = (crop.width * scaleX).roundToInt().coerceIn(1, width - x)
        val h = (crop.height * scaleY).roundToInt().coerceIn(1, height - y)

        val turned = request.rotation == MonoRotation.QUARTER_TURN ||
            request.rotation == MonoRotation.THREE_QUARTER_TURNS
        // What the region must measure *before* the rotation swaps its axes.
        val preWidth = if (turned) output.height else output.width
        val preHeight = if (turned) output.width else output.height

        val degrees = when (request.rotation) {
            MonoRotation.NONE -> 0f
            MonoRotation.QUARTER_TURN -> 90f
            MonoRotation.HALF_TURN -> 180f
            MonoRotation.THREE_QUARTER_TURNS -> 270f
        }

        // post* appends, so these apply in written order: scale, then rotate,
        // then mirror about the output's vertical axis. iOS composes the same
        // order, so a rotate-and-flip lands identically on both platforms.
        val matrix = Matrix()
        matrix.postScale(preWidth.toFloat() / w, preHeight.toFloat() / h)
        matrix.postRotate(degrees)
        if (request.flipHorizontal) matrix.postScale(-1f, 1f)

        var result = Bitmap.createBitmap(this, x, y, w, h, matrix, true)
        if (result != this) recycle()

        // Rounding inside createBitmap can land a pixel either side; the API
        // contract promises exact dimensions, so settle it here. Rarely runs.
        if (result.width != output.width || result.height != output.height) {
            val exact = Bitmap.createScaledBitmap(result, output.width, output.height, true)
            if (exact != result) result.recycle()
            result = exact
        }
        return result
    }
}

File: ios/monolens/Sources/monolens/ImageTransformer.swift
import ImageIO
import UIKit

/// Crop, rotate, flip, downscale and re-encode a still.
///
/// One decode and at most one redraw. The output size is worked out first, from
/// metadata alone, and the decode is then asked for only as many pixels as that
/// output needs — cropping a 12 MP photo down to a 1080 px square never
/// materialises the 12 MP bitmap. Cropping itself is free (`CGImage.cropping`
/// returns a view, not a copy), and the rotate/flip/scale that used to be three
/// separate full-size redraws are folded into a single composite draw, skipped
/// entirely when the crop already is the answer.
enum ImageTransformer {

  static func apply(_ request: ImageEditRequest) throws -> MediaInfo {
    let url = URL(fileURLWithPath: request.sourcePath)
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
      CGImageSourceGetCount(source) > 0
    else {
      throw MonolensError.decodeFailed("Could not read \(request.sourcePath)")
    }

    let full = try uprightSize(of: source)
    let crop = cropRect(request.crop, in: full)
    let output = outputSize(
      crop: crop, rotation: request.rotation, cap: request.maxDimension)

    // Ask the decoder for the smallest image that still covers the crop at the
    // output's resolution. Whole frame : kept region :: what we decode : output.
    let cropLongest = max(crop.width, crop.height)
    let fullLongest = max(full.width, full.height)
    let outputLongest = max(output.width, output.height)
    let wanted = cropLongest <= 0
      ? fullLongest
      : Int((Double(outputLongest) * Double(fullLongest) / Double(cropLongest)).rounded(.up))

    let decoded = try decodeUpright(source, maxPixelSize: min(fullLongest, wanted))

    // The decode is approximate, so rescale the crop into whatever came back
    // rather than assuming it honoured the request exactly.
    let cropped = try cut(decoded, to: crop, of: full)

    let needsRedraw =
      request.rotation != .none || request.flipHorizontal
      || cropped.width != output.width || cropped.height != output.height

    var image = needsRedraw
      ? try render(
        cropped, to: output, rotation: request.rotation,
        flipHorizontal: request.flipHorizontal, opaque: request.format == .jpeg)
      : cropped

    if !request.annotations.isEmpty {
      image = try annotate(image, with: request.annotations, size: output)
    }

    let byteSize = try encode(image, request: request)
    return MediaInfo(
      path: request.outputPath,
      width: Int64(image.width),
      height: Int64(image.height),
      durationMs: nil,
      byteSize: byteSize
    )
  }

  // MARK: - Geometry
  //
  // All of it in full-resolution upright pixels, so the output dimensions are
  // exact and independent of how much the decoder chose to downsample.

  private static func uprightSize(of source: CGImageSource) throws -> (
    width: Int, height: Int
  ) {
    guard
      let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let width = properties[kCGImagePropertyPixelWidth] as? Int,
      let height = properties[kCGImagePropertyPixelHeight] as? Int
    else { throw MonolensError.decodeFailed("No dimensions in image metadata") }

    let orientation = properties[kCGImagePropertyOrientation] as? UInt32 ?? 1
    let turned = orientation >= 5 && orientation <= 8
    return (turned ? height : width, turned ? width : height)
  }

  private static func cropRect(_ rect: NormalizedRect?, in full: (width: Int, height: Int))
    -> (x: Int, y: Int, width: Int, height: Int)
  {
    guard let rect else { return (0, 0, full.width, full.height) }
    let x = min(max(0, Int((rect.left * Double(full.width)).rounded())), full.width - 1)
    let y = min(max(0, Int((rect.top * Double(full.height)).rounded())), full.height - 1)
    let width = min(max(1, Int((rect.width * Double(full.width)).rounded())), full.width - x)
    let height = min(max(1, Int((rect.height * Double(full.height)).rounded())), full.height - y)
    return (x, y, width, height)
  }

  private static func outputSize(
    crop: (x: Int, y: Int, width: Int, height: Int), rotation: MonoRotation, cap: Int64?
  ) -> (width: Int, height: Int) {
    let turned = rotation == .quarterTurn || rotation == .threeQuarterTurns
    var width = turned ? crop.height : crop.width
    var height = turned ? crop.width : crop.height

    if let cap = cap.map(Int.init), cap >= 1, max(width, height) > cap {
      let factor = Double(cap) / Double(max(width, height))
      width = max(1, Int((Double(width) * factor).rounded()))
      height = max(1, Int((Double(height) * factor).rounded()))
    }
    return (width, height)
  }

  // MARK: - Pixels

  /// Decodes at most [maxPixelSize] on the longest edge, with the EXIF
  /// orientation already baked in — one pass instead of a full decode followed
  /// by a full-size redraw to straighten it.
  private static func decodeUpright(_ source: CGImageSource, maxPixelSize: Int) throws
    -> CGImage
  {
    let options: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceCreateThumbnailWithTransform: true,
      kCGImageSourceShouldCacheImmediately: true,
      kCGImageSourceThumbnailMaxPixelSize: max(1, maxPixelSize),
    ]
    guard
      let image = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary)
    else { throw MonolensError.decodeFailed("Could not decode image") }
    return image
  }

  private static func cut(
    _ image: CGImage, to crop: (x: Int, y: Int, width: Int, height: Int),
    of full: (width: Int, height: Int)
  ) throws -> CGImage {
    if crop.x == 0, crop.y == 0, crop.width == full.width, crop.height == full.height {
      return image
    }
    let scaleX = Double(image.width) / Double(full.width)
    let scaleY = Double(image.height) / Double(full.height)
    let rect = CGRect(
      x: (Double(crop.x) * scaleX).rounded(.down),
      y: (Double(crop.y) * scaleY).rounded(.down),
      width: max(1, (Double(crop.width) * scaleX).rounded()),
      height: max(1, (Double(crop.height) * scaleY).rounded())
    ).intersection(CGRect(x: 0, y: 0, width: image.width, height: image.height))

    guard !rect.isNull, let cropped = image.cropping(to: rect) else {
      throw MonolensError.invalidRange("Crop lies outside the image")
    }
    return cropped
  }

  /// The single composite pass: rotation, mirroring and the final scale at once.
  private static func render(
    _ image: CGImage, to output: (width: Int, height: Int), rotation: MonoRotation,
    flipHorizontal: Bool, opaque: Bool
  ) throws -> CGImage {
    let radians: CGFloat
    switch rotation {
    case .none: radians = 0
    case .quarterTurn: radians = .pi / 2
    case .halfTurn: radians = .pi
    case .threeQuarterTurns: radians = 3 * .pi / 2
    }
    let turned = rotation == .quarterTurn || rotation == .threeQuarterTurns

    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = opaque
    let size = CGSize(width: output.width, height: output.height)

    // UIKit's context is y-down, so a positive rotation reads clockwise — the
    // direction MonoRotation names.
    //
    // The CTM applies right-to-left, so `flip` written *before* `rotate` is the
    // one applied *after* it: the mirror is about the output's vertical axis,
    // which is what "flip horizontally" means to someone looking at the result.
    // Android composes the same order.
    let rendered = UIGraphicsImageRenderer(size: size, format: format).image { context in
      let cgContext = context.cgContext
      cgContext.translateBy(x: size.width / 2, y: size.height / 2)
      if flipHorizontal { cgContext.scaleBy(x: -1, y: 1) }
      cgContext.rotate(by: radians)

      // Pre-rotation extent: swapped when the turn is a quarter, so the drawn
      // image lands exactly on the output box.
      let drawWidth = turned ? size.height : size.width
      let drawHeight = turned ? size.width : size.height
      UIImage(cgImage: image).draw(
        in: CGRect(
          x: -drawWidth / 2, y: -drawHeight / 2, width: drawWidth, height: drawHeight))
    }

    guard let result = rendered.cgImage else {
      throw MonolensError.encodeFailed("Render produced no bitmap")
    }
    return result
  }

  /// Blur first, then everything else on top.
  ///
  /// Blur samples the media, so it belongs to the picture; text, stickers and
  /// strokes are drawn over the result. That ordering also means a caption
  /// placed across a blurred face stays legible, which is what an author
  /// expects and the reverse of what a naive back-to-front pass would give.
  private static func annotate(
    _ image: CGImage, with annotations: [AnnotationSpec], size: (width: Int, height: Int)
  ) throws -> CGImage {
    let canvas = CGSize(width: size.width, height: size.height)
    var working = CIImage(cgImage: image)

    working = AnnotationRenderer.applyingBlur(
      to: working,
      mask: AnnotationRenderer.blurMask(annotations, size: canvas),
      radius: AnnotationRenderer.blurRadius(annotations, size: canvas))

    working = AnnotationRenderer.compositing(
      AnnotationRenderer.overlayLayer(annotations, size: canvas), over: working)

    let context = CIContext(options: [.useSoftwareRenderer: false])
    guard let rendered = context.createCGImage(working, from: working.extent) else {
      throw MonolensError.encodeFailed("Could not flatten annotations")
    }
    return rendered
  }

  /// Straight to a file through ImageIO — no UIImage round trip and no
  /// intermediate `Data` held in memory.
  private static func encode(_ image: CGImage, request: ImageEditRequest) throws -> Int64 {
    // The raw UTIs rather than `UTType`, which needs iOS 14 — these strings are
    // stable and keep the deployment target at 13.
    let type = request.format == .png ? "public.png" : "public.jpeg"
    let url = URL(fileURLWithPath: request.outputPath)
    try? FileManager.default.removeItem(at: url)

    guard
      let destination = CGImageDestinationCreateWithURL(
        url as CFURL, type as CFString, 1, nil)
    else { throw MonolensError.encodeFailed("Could not open \(request.outputPath)") }

    var options: [CFString: Any] = [:]
    if request.format == .jpeg {
      options[kCGImageDestinationLossyCompressionQuality] =
        min(max(Double(request.quality), 1), 100) / 100
    }
    CGImageDestinationAddImage(destination, image, options as CFDictionary)

    guard CGImageDestinationFinalize(destination) else {
      throw MonolensError.encodeFailed("Could not write \(request.outputPath)")
    }
    return MediaProbe.fileSize(at: request.outputPath)
  }
}
